### PR TITLE
Bazel fusesoc invocation

### PR DESCRIPTION
**Pull request by drewmacrae**
Fixes: #15070
by removing two sources of duplicate flags:
* Repeated cores-roots no longer generate duplicate flags.
* We no longer try to compile verilator to use 1 and 4 cores simultaneously.

### CODE REVIEW

File: hw/BUILD [hunk 1]
```diff
@@ -30,7 +30,6 @@ fusesoc_build(
         "//:cores",
     ],
     data = ["//hw/ip/otbn:all_files"],
-    flags = ["--verilator_options=--threads 1"],
```
Comment (milesdai) [inline hw/BUILD:33]: Ah nice! Thanks for finding this. Is it now building the Verilator model to run on 4 threads? If so, is it possible to change it to run with 1 thread in CI? We run 4 parallel models simultaneously in CI, expecting each model to take 1 thread.

File: rules/fusesoc.bzl [hunk 2]
```diff
@@ -36,15 +36,20 @@ def _fusesoc_build_impl(ctx):
...
+        if c.dirname not in cores_roots:
+            cores_roots[c.dirname] = True
+
     # Note: the `fileset_top` flag used above is specific to the OpenTitan
     # project to select the correct RTL fileset.
     ctx.actions.run(
         mnemonic = "FuseSoC",
         outputs = outputs,
         inputs = ctx.files.srcs + ctx.files.cores,
         arguments = [
```
Comment (milesdai) [inline rules/fusesoc.bzl:50]: Not to pile on too much to this PR, but would you mind refactoring the argument passing to use Bazel's [Args object](https://bazel.build/rules/lib/Args)?

**Review by milesdai — APPROVED** [review]
Thanks for cleaning this up!

File: hw/BUILD (hunk 1 above)
Comment (drewmacrae) [reply hw/BUILD:33]: Here's where it's set in CI: https://cs.opensource.google/opentitan/opentitan/+/master:ci/scripts/run-verilator-tests.sh;drc=3f7956ee10c794d7327483d1172d22e80937ec72;l=17 I think, there's some unexercised code here that had me confused: https://cs.opensource.google/opentitan/opentitan/+/master:ci/scripts/build-chip-verilator.sh

Comment (milesdai) [reply hw/BUILD:33]: Ah I see -- okay looks good then.

File: rules/fusesoc.bzl (hunk 2 above)
Comment (drewmacrae) [reply rules/fusesoc.bzl:50]: I'll take a look

File: rules/fusesoc.bzl [hunk 3]
```diff
@@ -36,15 +36,20 @@ def _fusesoc_build_impl(ctx):
         verilator_options = ctx.attr.verilator_options[BuildSettingInfo].value
         flags.append("--verilator_options={}".format(" ".join(verilator_options)))
 
+    cores_roots = {}
+    for c in ctx.files.cores:
+        if c.dirname not in cores_roots:
+            cores_roots[c.dirname] = True
+
```
Comment (a-will) [inline rules/fusesoc.bzl:43]: nit: A `set` might be the data type that more directly expresses the intention here:
```suggestion
    cores_roots = set()
    for c in ctx.files.cores:
        cores_roots.add(c.dirname)
```

or
```python
    cores_roots = set([c.dirname for c in ctx.files.cores])
```

**Review by a-will — APPROVED** [review]
Some things to think about (which you might already have the answers to):

Is there evidence somewhere that the verilated model is substantially faster with 4 threads? Have we tested this?

Also, is the remote cache that CI uses only intended for CI as a consumer? Shouldn't we also at least build this default in CI? (I also feel a bit iffy about using a substantially different default build configuration for CI vs. local users, since we're not testing the latter. There may be unexpected side effects for timing.)

**Comment by drewmacrae** [comment]
> Is there evidence somewhere that the verilated model is substantially faster with 4 threads? Have we tested this?

Yes, it runs about twice as fast with 4 threads, and another 30% faster with 8 IIRC, The default of 4 isn't changed in this PR, but felt like a good balance. I sometimes run with more when I anticipate long Verilated test runs, but changing it does invalidate my cache. @milesdai did some more benchmarking and also sets it to 1-thread for CI where we parallelize the tests by running multiples anyway.

In my .bazelrc file I have 
```
#build verilator for 8 threads instead of 4 for a ~30% speedup
build:fast --//hw:verilator_options=--threads,8
build:fast --local_test_jobs=HOST_CPUS*0.11
```
that lets me use the argument `--config=fast` to build and test single verilator tests more effectively.

**Comment by drewmacrae** [comment]
> Also, is the remote cache that CI uses only intended for CI as a consumer? Shouldn't we also at least build this default in CI? (I also feel a bit iffy about using a substantially different default build configuration for CI vs. local users, since we're not testing the latter. There may be unexpected side effects for timing.)

Regarding the cache,  we structure the keys differently in CI than in local builds so you're not going to see hits at present unfortunately. Regarding timing and the build, I'm reasonably comfortable trusting verilator code to compile the verilated model for different number of threads when we're sensitive enough to performance that we're fielding complaints and haven't yet seen an issue. I think you're right that going forward we should have a low threshold for recommending multithreaded verilator if we have trouble reproducing issues that appear in CI.

File: rules/fusesoc.bzl (hunk 3 above)
Comment (drewmacrae) [reply rules/fusesoc.bzl:43]: Huh... I was today years old when I learned that "set" was built into Python, but I don't think they exist in Starlark so I wouldn't expect this to work.

Comment (a-will) [reply rules/fusesoc.bzl:43]: Oh yeah... It's Starlark. Doh, I forgot! I can't decide whether to be bothered by the "sort of Python" nature of this language. But in any case, withdrawn!

File: rules/fusesoc.bzl (hunk 2 above)
Comment (drewmacrae) [reply rules/fusesoc.bzl:50]: I've done this in #15554 Should I merge this one and rebase it or would you prefer to wait for it all at once?

Comment (milesdai) [reply rules/fusesoc.bzl:50]: I'm fine with merging them separately

**Pull request merged**